src/compiler/checker.ts

```diff
@@ -5000,42 +5000,78 @@ namespace ts {
             // * className.prototype.method = expr
             else if (isBinaryExpression(decl) ||
                 isPropertyAccessExpression(decl) && isBinaryExpression(decl.parent)) {
-                return getJSInitializerType(decl, symbol, getAssignedJavascriptInitializer(isBinaryExpression(decl) ? decl.left : decl)) ||
+                return getJSObjectLiteralInitializerType(decl, symbol, getAssignedJavascriptInitializer(isBinaryExpression(decl) ? decl.left : decl)) ||
                     getWidenedTypeFromJSSpecialPropertyDeclarations(symbol);
             }
             else if (isParameter(decl)
                 || isPropertyDeclaration(decl)
                 || isPropertySignature(decl)
                 || isVariableDeclaration(decl)
                 || isBindingElement(decl)) {
-                // Use type from type annotation if one is present
+                // In order, use
+                // 1. type from type annotation if one is present
+                // 2. Type of js container, if the initializer is an object literal
+                // 3. Type of js container intersected with initializer, if function expression, class expression, or IIFE.
                 const isOptional = isParameter(decl) && isJSDocOptionalParameter(decl) ||
                     !isBindingElement(decl) && !isVariableDeclaration(decl) && !!decl.questionToken;
                 const declaredType = tryGetTypeFromEffectiveTypeNode(decl);
-                return declaredType && addOptionality(declaredType, isOptional) ||
-                    getJSInitializerType(decl, symbol, getDeclaredJavascriptInitializer(decl)) ||
-                    getWidenedTypeForVariableLikeDeclaration(decl, /*includeOptionality*/ true);
+                if (declaredType) {
+                    return addOptionality(declaredType, isOptional);
+                }
+                const init = getDeclaredJavascriptInitializer(decl);
+                const initializerType = getJSObjectLiteralInitializerType(decl, symbol, init);
+                if (initializerType) {
+                    return initializerType;
+                }
+                const widened = getWidenedTypeForVariableLikeDeclaration(decl, /*includeOptionality*/ true);
+                if (init && getJavascriptInitializer(init, isPrototypeAccess(decl.name))) {
+                    if (symbol.flags & SymbolFlags.Transient && hasEntries(symbol.exports) && widened.flags & TypeFlags.StructuredType) {
+                        resolveStructuredTypeMembers(widened as StructuredType);
+                        const exports = createSymbolTable();
+                        forEachDeclarationInChain(decl, s => {

```

Comment:
Does this need to be done up-front, or can it be deferred until members are requested in `resolveStructuredTypeMembers` on the result of this call? The `resolveStructuredTypeMembers` on the other intersection member (to look for is members to check for things to merge) above leads me to believe that it may be good to do so. AFAIK, we'd need to set a `CheckFlag` or something on this intersection's symbol (or maybe its anonymous object symbol?) and then move this work into `resolveStructuredTypeMembers`.